router/src/components/router.rs

```diff
@@ -57,14 +57,29 @@ pub fn Router(
     let navigate = use_navigate();
     let navigate = SendWrapper::new(navigate);
     let router_hook = Box::new(move |path: &str| {
-        let path = path.to_string();
-        // delay by a tick here, so that the Action updates *before* the redirect
-        request_animation_frame({
-            let navigate = navigate.clone();
-            move || {
-                navigate(&path, Default::default());
+        match Url::try_from(path) {
+            Ok(url) => {
+                let current_origin = leptos_dom::helpers::location()
+                    .origin()
+                    .unwrap_or_default();
+                if url.origin == current_origin {
```

Comment:
This check fails for local URLs like `/foobar` because in the browser, `Url::try_from` inserts a fake `http://leptos` base:

https://github.com/leptos-rs/leptos/blob/a903e19eb24371e59b6c36dfdddc2ca1940866b0/router/src/history/url.rs#L47-L57

This is simply because the `new URL()` constructor will fail with an incomplete URL
```js
new URL("/foobar");
// throws `Uncaught TypeError: URL constructor: /foobar is not a valid URL.`
```

I *think* it's safe to add a `url.origin = current_origin || url.origin == "http://leptos"`, maybe refactoring these two places to share a constant, or adding a helper method to `Url` like `is_app_origin`.

Reply:
Why does this have a double slash here?
https://github.com/leptos-rs/leptos/blob/a903e19eb24371e59b6c36dfdddc2ca1940866b0/router/src/history/url.rs#L48

Reply:
What about the server? Would it also fail there (but doesn't, because that code isn't run on the server)?

The client-side code for `Url::try_from()` doesn't seem quite correct: I think we should apply the rules of https://datatracker.ietf.org/doc/html/rfc3986#section-4, i.e., treat the input as an URI-reference and, if it's a relative reference, resolve it against a base URI (probably the current window location). Eexamples in section 5.4.1.

Reply:
The only place that 

> Why does this have a double slash here?
> 
> https://github.com/leptos-rs/leptos/blob/a903e19eb24371e59b6c36dfdddc2ca1940866b0/router/src/history/url.rs#L48

#1764 describes an issue in which using a route like `http://my-app.com///foo` causes an error, instead of matching the path `//foo`. Using something like `new URL("//foo", window.location.origin)` causes the browser to misinterpret this as the URL `"https://foo/"`. So in that case, we manually build the whole URL from `window.location.origin`. In other cases, we just use `new URL()` with a fake base.

If you fire up the browser console it might make more sense: observe the difference in behavior between these two:
```js
// this is fine
new URL("/foo", window.location.origin);
// this is not
new URL("//foo", window.location.origin);
```

Hence the check that's currently in place.

It would be possible to change this by always getting the `window.location.origin` and doing string concatenation, but note you'll need to check for a `/` at the beginning, as the following set also do not behavior properly
```js
// good
new URL(window.location.origin + "/foo")
// good
new URL(window.location.origin + "//foo")
// bad
new URL(window.location.origin + "foo")
```

> The client-side code for `Url::try_from()` doesn't seem quite correct: I think we should apply the rules of https://datatracker.ietf.org/doc/html/rfc3986#section-4, i.e., treat the input as an URI-reference and resolve it against a base URI (probably the current window location). Examples in section 5.4.1.

I might be misunderstanding you, but as far as I can tell neither the `url` crate nor the `new URL()` constructor in the browser do this kind of resolution; they simply require complete URLs, hence the workaround. I don't think it makes sense to implement the kind of resolution described in 5.4.1 in the context of our router.

Reply:
> Using something like `new URL("//foo", window.location.origin)` causes the browser to misinterpret this as the URL `"https://foo/"`.

Why "misinterpret"?  According to the RFC, `//foo` is a network-path reference and it's correct behavior to resolve it as `http://foo` or `https://foo` (depending on whether the current origin uses `http` or `https`).

Try the following with JS/WASM disabled. The browser correctly links / redirects to `http://example.com`.

```rust
use leptos::*;
use leptos_meta::*;
use leptos_router::*;

#[component]
pub fn App() -> impl IntoView {
    provide_meta_context();

    let redirect = create_server_action::<Redirect>();
    
    view! {
        <Stylesheet id="leptos" href="/pkg/leptos-playground.css"/>
        <Title text="Leptos Playground"/>

        <a href="//example.com">Click me</a>
        <ActionForm action=redirect>
            <button type="submit">Redirect</button>
        </ActionForm>
    }
}

#[server(Redirect)]
async fn redirect() -> Result<(), ServerFnError> {
    leptos_axum::redirect("//example.com");

    Ok(())
}
```

Reply:
> as far as I can tell neither the `url` crate nor the `new URL()` constructor in the browser do this kind of resolution

The `new URL()` constructor, when passing a base URI as 2nd argument, seems to do the resolution specified by https://datatracker.ietf.org/doc/html/rfc3986#section-4